File: Game/Scene.cs
using OpenTK.Graphics.OpenGL4;
using OpenTK.Mathematics;
using Game.Rendering;

namespace Game;

public class Scene
{
  private readonly Shader _shader = Shader.Load(ShaderDefault.Vertex, ShaderDefault.Fragment);
  private float _width;
  private float _height;
  private float _scale;

  private Scene()
  {
    GL.Enable(EnableCap.Blend);
    GL.BlendFunc(BlendingFactor.SrcAlpha, BlendingFactor.OneMinusSrcAlpha);
    GL.Enable(EnableCap.DepthTest);
    GL.DepthMask(true);
    GL.DepthFunc(DepthFunction.Lequal);
    GL.DepthRange(-1, 1);

    GL.BindVertexArray(GL.GenVertexArray());
  }

  public static Scene Create(float width, float height, float scale)
  {
    var scene = new Scene();
    scene._shader.Dimensions(width, height, scale);
    scene._width = width;
    scene._height = height;
    scene._scale = scale;
    return scene;
  }

  public void Clear()
  {
    GL.ClearColor(1f, 0.5f, 0.5f, 1f);
    GL.ClearDepth(1);
    GL.Clear(ClearBufferMask.ColorBufferBit | ClearBufferMask.DepthBufferBit);
  }

  public Actor CreateActor() =>
    Actor.Create(_shader, new Vector3(_width, _height, _scale));

  private void Camera(float x, float y)
   => _shader.Camera(x, y);

    public void ShakeCameraRandomly(float range)
  {
    var random = new Random();
    var between = (range + range + 1);
    var x = (float) random.NextDouble() * between - range;
    var y = (float) random.NextDouble() * between - range;
    Camera(x, y);
  }
}

File: Blazor/wwwroot/Rendering/Scene.js
class Scene {
    #shader
    #width
    #height
    #scale
    #camera

    constructor() {
        this.#camera = {x: 0, y: 0}

        gl.enable(gl.BLEND);
        gl.blendFunc(gl.SRC_ALPHA, gl.ONE_MINUS_SRC_ALPHA)
        gl.enable(gl.DEPTH_TEST)
        gl.depthMask(gl.FALSE)
        gl.depthFunc(gl.LEQUAL)
        gl.depthRange(-1, 1)
    }

    static create(width, height, scale) {
        let scene = new Scene()
        scene.#shader = Shader.load(ShaderDefault.vertex, ShaderDefault.fragment);
        scene.#width = width
        scene.#height = height
        scene.#scale = scale
        scene.#shader.dimensions(width, height, scale)

        return scene
    }

    clear() {
        gl.clearColor(1.0, 0.5, 0.5, 1.0)
        gl.clearDepth(1.0)
        gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT)
    }

    createActor() {
        return new Actor(this.#shader)
    }

    camera(x, y) {
        this.#camera = {x, y}
        this.#shader.camera(this.#camera.x, this.#camera.y)
    }


    dimensions(width, height, scale) {
        this.#shader.dimensions(width, height, scale)
    }
}